fix(connector): Use `ConnectorError::InvalidConnectorConfig` for an invalid `CoinbaseConnectorMeta`

bkontur opened the pull request:
## Type of Change
<!-- Put an `x` in the boxes that apply -->

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
<!-- Describe your changes in detail -->

I've removed `to_connector_meta_from_secret_with_required_field`, which uses `MissingRequiredField` as an error (I also didn't like the unnecessary initialization of the `MissingRequiredField` before we knew if it would be needed). Instead, I added a `TryFrom` implementation for `CoinbaseConnectorMeta`, which uses the requested `ConnectorError::InvalidConnectorConfig`. Additionally, I included `CoinbaseConnectorMeta::try_from` in the `validate_auth_and_metadata_type` as suggested.

### Migration

```
-- The migration is for adding an empty `pricing_type` attribute for `CoinbaseConnectorMeta` where it is missing.
UPDATE merchant_connector_account
SET metadata = jsonb_insert(
    metadata,
    '{pricing_type}',
    '""',
    true
)
WHERE connector_name = 'coinbase'
    AND NOT metadata ? 'pricing_type';
```

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables


## Motivation and Context
<!--
Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.

If you don't have an issue, we'd recommend starting with one first so the PR
can focus on the implementation (unless it is an obvious bug or documentation fix
that will have little conversation).
-->
This PR addresses the bug from the issue: https://github.com/juspay/hyperswitch/issues/2899.

## How did you test it?
<!--
Did you write an integration/unit/API test to verify the code changes?
Or did you test this change manually (provide relevant screenshots)?
-->

**Run unit test:**
```
cargo test --release  --package router --lib connector::coinbase::transformers::tests::coinbase_payments_request_try_from_works -- --exact
```
![image](https://github.com/juspay/hyperswitch/assets/8159517/a296af9b-a454-4d9b-a8ad-1939a837f04b)


### Migration script test

https://onecompiler.com/postgresql/3zz2tpkad

**Test script with test data and run migration:**
```label=dsa
CREATE TABLE merchant_connector_account (
    id SERIAL PRIMARY KEY,
    connector_name VARCHAR(255) NOT NULL
);
ALTER TABLE merchant_connector_account ADD COLUMN metadata JSONB DEFAULT NULL;

-- insert test data
INSERT INTO merchant_connector_account (
        id,
        connector_name,
        metadata
    )
VALUES
  (1, 'coinbase', '{ "pricing_type": "fixed-rate" }'),
  (2, 'coinbase', '{ "whatever_attribute": "abc" }'),
  (3, 'coinbase', '{}');

-- Before
select * from merchant_connector_account;

-- Run migration:
-- The migration is for adding an empty `pricing_type` attribute for `CoinbaseConnectorMeta` where it is missing.
UPDATE merchant_connector_account
SET metadata = jsonb_insert(
    metadata,
    '{pricing_type}',
    '""',
    true
)
WHERE connector_name = 'coinbase'
    AND NOT metadata ? 'pricing_type';
    
-- After
select * from merchant_connector_account;
```

**Output:**
```
Output:

CREATE TABLE
ALTER TABLE
INSERT 0 3
 id | connector_name |            metadata            
----+----------------+--------------------------------
  1 | coinbase       | {"pricing_type": "fixed-rate"}
  2 | coinbase       | {"whatever_attribute": "abc"}
  3 | coinbase       | {}
(3 rows)

UPDATE 2
 id | connector_name |                     metadata                      
----+----------------+---------------------------------------------------
  1 | coinbase       | {"pricing_type": "fixed-rate"}
  2 | coinbase       | {"pricing_type": "", "whatever_attribute": "abc"}
  3 | coinbase       | {"pricing_type": ""}
(3 rows)
```

**Run locally:**

**Invalid metadata:**
![image](https://github.com/juspay/hyperswitch/assets/8159517/5e087dad-9dab-4441-a444-787d8e720c1a)

**Empty metadata:**
![image](https://github.com/juspay/hyperswitch/assets/8159517/0f745195-0b0b-4276-aeba-c18ad11d3749)

**Correct metadata:**
![image](https://github.com/juspay/hyperswitch/assets/8159517/a5fe8ffa-12d2-449d-8363-a214aadca2fb)



## Checklist
<!-- Put an `x` in the boxes that apply -->

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [x] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

srujanchikke commented on crates/router/src/connector/coinbase/transformers.rs:
```diff
@@ -250,6 +252,17 @@ pub struct CoinbaseConnectorMeta {
     pub pricing_type: String,
 }
 
+impl TryFrom<&Option<pii::SecretSerdeValue>> for CoinbaseConnectorMeta {
```
This changes will impact previous merchants who are using coinbase, This PR needs data migrations. Could you please refer https://github.com/juspay/hyperswitch/pull/2746 description for more info.

bkontur replied:
[Issue#2899](https://github.com/juspay/hyperswitch/issues/2899) does not mention any data migrations.
I've checked that #2746, and in that case, the migration involved replacing `terminalId` with `terminal_id` due to the removal of `#[serde(rename_all = "camelCase")]` [here](https://github.com/juspay/hyperswitch/pull/2746/files#diff-bd16bd6d296808eacd72cdeb8a8bfbe54bc788b30b844301e136c4396a53eeceL421-R425).

However, my issue is different. If I understand correctly, this issue is about empty JSON for `CoinbaseConnectorMeta` or the absence of the `pricing_type` attribute in the JSON, and the goal is to add validations to prevent storing invalid `CoinbaseConnectorMeta`.

So, I am not sure about the migration here. If empty JSON or JSON without `pricing_type` is stored, how should the migration to the correct `CoinbaseConnectorMeta` be handled? What value should be set for `pricing_type`?

srujanchikke replied:
No worries, We will take this up as these are data migrations. FYI, I was talking about existing merchants, we need to add `pricing_type` in metadata and set it to empty string. In that case, when merchant tries to make payment connector throws error and they are expected to update the metadata.

bkontur replied:
@srujanchikke aha, ok, thank you, make senses now, so I prepared migration script and simple test demonstration that it works, please see PR description.

srujanchikke reviewed (changes requested):
Please attach screenshots of testing if possible. refer https://github.com/juspay/hyperswitch/pull/2746 to know how to test this issue.

bkontur commented on crates/router/src/configs/settings.rs:
```diff
@@ -556,7 +556,7 @@ impl From<Database> for storage_impl::config::Database {
             pool_size: val.pool_size,
             connection_timeout: val.connection_timeout,
-            queue_strategy: val.queue_strategy.into(),
+            queue_strategy: val.queue_strategy,
```
This was already fixed with other PR: https://github.com/juspay/hyperswitch/pull/3231/files#diff-352c4b9115d187dd2f0e24e6a057b11959bd781a38ac9d26f905983b9517bbd6L559-R559

bkontur commented on crates/router/src/connector/utils.rs:
```diff
@@ -785,7 +785,7 @@ impl CardData for api::Card {
         Secret::new(format!(
             "{}{}{}",
-            self.card_exp_month.peek().clone(),
+            self.card_exp_month.peek(),
```
Reverted and moved to the separate PR: https://github.com/juspay/hyperswitch/pull/3259

bkontur commented on crates/router/src/connector/worldline/transformers.rs:
```diff
@@ -342,7 +342,7 @@ fn make_card_request(
     ccard: &payments::Card,
 ) -> Result<CardPaymentMethod, error_stack::Report<errors::ConnectorError>> {
-    let expiry_year = ccard.card_exp_year.peek().clone();
+    let expiry_year = ccard.card_exp_year.peek();
```
Reverted and moved to the separate PR: https://github.com/juspay/hyperswitch/pull/3259

bkontur commented on crates/router/src/core/fraud_check.rs:
```diff
@@ -167,10 +167,9 @@ where
                         Some(frm_configs_value) => {
                             let frm_configs_struct: Vec<FrmConfigs> = frm_configs_value
-                                .iter()
+                                .into_iter()
```
Reverted and moved to the separate PR: https://github.com/juspay/hyperswitch/pull/3259

bkontur commented on crates/router/src/core/payment_methods/vault.rs:
```diff
@@ -53,7 +53,7 @@ impl Vaultable for api::Card {
             name_on_card: self
                 .card_holder_name
-                .clone()
+                .as_ref()
```
Reverted and moved to the separate PR: https://github.com/juspay/hyperswitch/pull/3259

bkontur commented on crates/router/src/connector/coinbase/transformers.rs:
```diff
@@ -419,3 +428,159 @@ pub struct TimelinePayment {
 }
+
+#[cfg(test)]
+mod tests {
```
Ok, thank you. I have moved that test as suggested.

bkontur commented:
> Please attach screenshots of testing if possible. refer #2746 to know how to test this issue.

Ok, I will try the next week

srujanchikke reviewed (changes requested):
Also need to make changes where ever `CoinbaseConnectorMeta` is used
https://github.com/juspay/hyperswitch/blob/ecf51b5e3a30f055634edfafcd36f64cef535a53/crates/router/src/connector/coinbase/transformers.rs#L263

`let metadata = item.router_data.get_connector_meta()?;
let connector_meta: CoinbaseConnectorMeta =
        metadata.parse_value(“CoinbaseConnectorMeta”)
            .change_context(errors::ConnectorError::InvalidConnectorConfig {
                config: “Merchant connector account metadata”,
            })?;`

bkontur commented on crates/router/src/connector/coinbase/transformers.rs:
```diff
@@ -250,6 +252,17 @@ pub struct CoinbaseConnectorMeta {
+    fn try_from(meta_data: &Option<pii::SecretSerdeValue>) -> Result<Self, Self::Error> {
+        utils::to_connector_meta_from_secret(meta_data.clone()).change_context(
+            errors::ConnectorError::InvalidConnectorConfig {
+                config: "`pricing_type` not present in `CoinbaseConnectorMeta`",
```
ok, changed, https://github.com/juspay/hyperswitch/pull/3168/commits/0bc067471d6c924a55df7a050ab7f344bb8ced28

bkontur commented:
> > Please attach screenshots of testing if possible. refer #2746 to know how to test this issue.
> 
> Ok, I will try the next week

@srujanchikke testing done with local `docker compose --file docker-compose-development.yml up -d`.
I like the prepared postman stuff :), please, see the description with attached screenshots.

bkontur commented:
@srujanchikke I think ,I've addressed all your review comments, added migration script (with simple test for sql) and added screenshots from postman testing, anything else?

srujanchikke commented on crates/router/tests/connectors/coinbase.rs:
```diff
@@ -1,6 +1,13 @@
+use std::marker::PhantomData;
```
Can you revert this test file changes, Currently we are adding non critical testcases using [new man postman ](https://github.com/juspay/hyperswitch/wiki/A-better-storage-format-for-postman-collections). It will be quite effort for you to follow this, Instead we can internally assign this to someone from team.

srujanchikke reviewed (changes requested):
Hi @bkontur ,
Apologies for the delay, Given some minor comments, Please resolve it.

bkontur commented on crates/router/src/connector/coinbase/transformers.rs:
```diff
@@ -260,11 +270,7 @@ fn get_crypto_specific_payment_data(
-            item.connector_meta_data.clone(),
-            "Pricing Type Not present in connector meta data",
-        )?;
+    let connector_meta = CoinbaseConnectorMeta::try_from(&item.connector_meta_data)?;
```
ok, got it, fixed
